modules/alert/mail/main.c

```diff
@@ -0,0 +1,197 @@
+
+	/* https://curl.haxx.se/libcurl/c/smtp-ssl.html */
+	if (curl) {
+		/* Set username and password */
+		curl_easy_setopt(curl, CURLOPT_USERNAME, state->username);
+		curl_easy_setopt(curl, CURLOPT_PASSWORD, state->password);
+
+		/* SSL based connection */
+		char server[128];
+		sprintf(server, "smtps://%s", state->server);
+		curl_easy_setopt(curl, CURLOPT_URL, server);
+
+#ifdef SKIP_PEER_VERIFICATION
+		curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0L);
+#endif
+
+#ifdef SKIP_HOSTNAME_VERIFICATION
+		curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0L);
+#endif
+
+		/* Mail sender */
+		curl_easy_setopt(curl, CURLOPT_MAIL_FROM, FROM);
+
+		/* Mail recipients */
+		while ((recipient = strsep(&state->recipients, ",")) != NULL)
+			recipients = curl_slist_append(recipients, recipient);
+		curl_easy_setopt(curl, CURLOPT_MAIL_RCPT, recipients);
+
+		/* Message body */
+		char body[2048];
+		sprintf(body, "Subject: [Haka] alert: %s%s\r\n",
+			update ? "update " : "", alert_tostring(id, time, alert, "", " ", false));
+
+		email_ctx.body = body;
+		email_ctx.size = strlen(body);
+
+		curl_easy_setopt(curl, CURLOPT_READFUNCTION, payload_source);
+		curl_easy_setopt(curl, CURLOPT_READDATA, &email_ctx);
+		curl_easy_setopt(curl, CURLOPT_UPLOAD, 1L);
+
+#ifdef ENABLE_VERBOSE
+		curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
+#endif
+		/* Send the message */
+		res = curl_easy_perform(curl);
+
+		/* Check for errors */
+		if (res != CURLE_OK) {
+			error("curl_easy_perform() failed: %s\n",
+				curl_easy_strerror(res));
+			return false;
+		}
+	}
+	return true;
+}
+
+static bool do_alert(struct alerter *state, uint64 id, const struct time *time, const struct alert *alert)
+{
+	return send_mail((struct mail_alerter *)state, id, time, alert, false);
+}
+
+static bool do_alert_update(struct alerter *state, uint64 id, const struct time *time, const struct alert *alert)
+{
+	return send_mail((struct mail_alerter *)state, id, time, alert, true);
+}
+
+struct alerter_module *init_alerter(struct parameters *args)
+{
+	struct mail_alerter *mail_alerter = malloc(sizeof(struct mail_alerter));
+	if (!mail_alerter) {
+		error("memory error");
+		return NULL;
+	}
+
+	mail_alerter->module.alerter.alert = do_alert;
+	mail_alerter->module.alerter.update = do_alert_update;
+
+	/* Mandatory fields: SMTP server, credentials and mail recipient */
+	const char *server = parameters_get_string(args, "mail_server", NULL);
+	const char *username = parameters_get_string(args, "mail_username", NULL);
+	const char *password = parameters_get_string(args, "mail_password", NULL);
+	const char *recipients = parameters_get_string(args, "mail_recipient", NULL);
+	if (!server || !username || !password || !recipients) {
+	        error("missing mandatory field");
+		free(mail_alerter);
+	        return NULL;
+	}
+
+	mail_alerter->server = strdup(server);
+	mail_alerter->username = strdup(username);
+	mail_alerter->password = strdup(password);
+	mail_alerter->recipients = strdup(recipients);
```

Comment:
We should process recipients here and store the curl_slist in `mail_alerter`. This avoid the bug I commented on line 101. It save some computation, and avoid the `recipients` global var.